### PR TITLE
Reconcile AFS consumed usage for stopped LocalQueues

### LINKED ISSUE #15383: 🐛 Admission Fair Sharing undercounts usage after LocalQueue Hold
**What happened**:
When a LocalQueue is put on Hold, its Admission Fair Sharing (AFS) usage history stops updating even though admitted workloads keep running. If those workloads finish before the queue resumes, their usage during Hold is missed.

I reproduced this with two Jobs on Kind. Both requested 1 CPU and ran `sleep 180`. After putting one queue on Hold, letting both Jobs finish, and resuming it, `consumedResources.cpu` was roughly 469m for the active queue and 7m for the queue put on Hold. 

This could let new work from the resumed queue go ahead of work from other queues.

**What you expected to happen**:
Both queues ran the same work, so they should have similar usage histories after resume.

**How to reproduce it (as minimally and precisely as possible)**:
With AdmissionFairSharing enabled:

1. Set the following in the configuration (`managerConfig.controllerManagerConfigYaml`):

   ```yaml
   admissionFairSharing:
     usageHalfLifeTime: 30s
     usageSamplingInterval: 2s
   ```

    (The short configuration settings help the test finish in a few minutes)

2. Create two LocalQueues, `open` and `held`, both with weight 1, under one ClusterQueue. Set `spec.admissionScope.admissionMode: UsageBasedAdmissionFairSharing` and give it 4 CPUs of quota.
3. Submit one Job to each queue, each requesting 1 CPU and running `sleep 180`.
4. Wait 30 seconds after both jobs are admitted, then set `held`'s `spec.stopPolicy` to `Hold`.
5. Let both Jobs finish, wait another 30 seconds, then set `held`'s stop policy to `None`.
6. Compare `consumedResources.cpu` and `lastUpdate` for each LocalQueue's `status.fairSharing.admissionFairSharingStatus`.

During Hold, the queue's `consumedResources.cpu` and `lastUpdate` stayed frozen while its Job kept running. After the Job finished, resuming the queue dropped its `consumedResources.cpu` to **7m**, compared with **469m** for the active queue. Kueue decayed the old history as though the entire pause had been idle.

**Anything else we need to know?**:
In [`LocalQueueReconciler.Reconcile`](https://github.com/kubernetes-sigs/kueue/blob/730f57dd18194b6f0825e6231301ac9d3e9b7d77/pkg/controller/core/localqueue_controller.go), a stopped queue causes an early return before `reconcileConsumedUsage` updates AFS history.

If you agree that usage should still count during Hold, I’d be happy to work on a fix.

**Environment**:
- Kubernetes: v1.36.1 on Kind, linux/arm64
- Kueue: v0.20.0-devel-586-g730f57dd1, locally built and installed with Helm
- Full commit: 730f57dd18194b6f0825e6231301ac9d3e9b7d77

AI assistance was used during the investigation and drafting of this report.

### PR DESCRIPTION
**Pull request by shashankvarma499**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it?

When a LocalQueue's `StopPolicy` is set to `Hold` (or `HoldAndDrain`), the
reconciler returned early, before reconciling its Admission Fair Sharing (AFS)
consumed usage. Admitted Workloads that kept running during the hold therefore
had their usage decay as though the queue were idle, and the queue's usage
history stopped being sampled because no requeue was scheduled.

This change makes the reconciler keep reconciling AFS consumed usage for
stopped queues, so running Workloads continue contributing to the queue's
consumption history while it is held.

Fixes #15383

#### Useful notes for your reviewer

The AFS reconciliation is extracted into a `reconcileAfsUsage` helper so it can
run for both active and stopped queues. The helper reports whether a sampling
tick actually ran, so the active path still only rebuilds the ClusterQueue when
usage was actually updated.

I used AI assistance during the investigation and drafting of this PR.

#### Release note

```release-note
Admission Fair Sharing: a LocalQueue that is put on Hold now keeps accumulating usage from already-admitted Workloads, so its consumption history no longer decays as idle time while held.
```

## AI summary

- Fixed AFS usage reconciliation for LocalQueues with `Hold` or `HoldAndDrain`.
- Held queues now continue updating usage history for running admitted Workloads.
- Added unit and integration test coverage.

### Suggested release note

AFS: Fixed a bug that caused usage history to stop updating for held LocalQueues with running admitted Workloads. Kueue now continues tracking their consumed usage during the hold.

### CODE REVIEW

**Review by mimowo — COMMENTED** [review]
I think it would be great to have an integration test for the bugfix

**Comment by shashankvarma499** [comment]
Makes sense. I'll add an integration test that drives the held-LocalQueue path so the AFS sampling is exercised end to end, not just the unit-level helper. Will push it shortly.

**Comment by shashankvarma499** [comment]
Added an integration test covering the held-queue path: admit a Workload so AFS usage accrues, set the LocalQueue to Hold, then assert a sampling tick still runs while admitted Workloads keep running (LastUpdate advances and ConsumedResources does not decay as idle).

Verified locally: the new spec passes under envtest

```
KUBEBUILDER_ASSETS=... bin/ginkgo --focus='keeps reconciling AFS consumed usage while the LocalQueue is held' ./test/integration/singlecluster/scheduler/fairsharing
Ran 1 of 48 Specs in 10.697 seconds — 1 Passed
```

`make ci-lint` is clean.

**Comment by mimowo** [comment]
cc @apullo777 could you please give it a pass?

**Comment by shashankvarma499** [comment]
Added an integration test for the held-queue path in `test/integration/singlecluster/scheduler/fairsharing/fair_sharing_test.go`: admit a Workload so AFS consumed usage accrues, set the LocalQueue StopPolicy to Hold, then assert that a sampling tick still runs (LastUpdate advances) and ConsumedResources does not decay as idle while the admitted Workload keeps running. Happy to reshape it if you had a different scenario in mind.

File: pkg/controller/core/localqueue_controller.go [hunk 1]
```diff
@@ -193,7 +193,21 @@ func (r *LocalQueueReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		if err != nil {
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
+		if afs.Enabled(r.admissionFSConfig) {
+			// A stopped LocalQueue no longer admits new Workloads, but already
+			// admitted Workloads keep running. Reconcile AFS consumed usage so
+			// their usage during Hold is still counted instead of decaying as
+			// idle time (#15383).
+			result, _, err := r.reconcileAfsUsage(ctx, &queueObj)
```
Comment (apullo777) [inline pkg/controller/core/localqueue_controller.go:204]: One edge case here: with LocalQueue metrics enabled, a held reconcile recreates the AFS usage series after Hold cleared it, and deleting the LocalQueue afterwards leaves that series behind.

I can reproduce the series count as 1 → 0 → 1 → 1 across active reconcile → Hold → held reconcile → Delete.

Could we either avoid reporting it for stopped queues, or make the delete path clean it up?

File: pkg/controller/core/localqueue_controller.go [hunk 2]
```diff
@@ -193,7 +193,21 @@ func (r *LocalQueueReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+			if err != nil {
+				return ctrl.Result{}, client.IgnoreNotFound(err)
```
Comment (apullo777) [inline pkg/controller/core/localqueue_controller.go:206]: Another edge case: a held LocalQueue whose ClusterQueue is missing now returns `ErrCqNotFound` from this path. Before this change, the stopped path returned successfully.

After the sampling interval passes, I see reconciles repeatedly fail until the ClusterQueue appears. Could we treat a missing ClusterQueue as "skip this sample" here, and add a held-queue test for it?

File: pkg/controller/core/localqueue_controller.go [hunk 3]
```diff
@@ -213,27 +227,42 @@ func (r *LocalQueueReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 
 	if afs.Enabled(r.admissionFSConfig) {
```
Comment (apullo777) [inline pkg/controller/core/localqueue_controller.go:229]: One related case I noticed: an inactive ClusterQueue returns just above this block, so `Hold` at the ClusterQueue level seems to have the same accounting gap when a running Workload finishes before the queue resumes.

I don't think this needs to block this PR, especially since the inactive-CQ path has a few more semantics to consider. Do you think it's worth covering here, or should we track it separately?

**Review by apullo777 — COMMENTED** [review]
Thanks for fixing this. The approach looks right to me!

One wording/test point: the undercount happens when admitted usage changes while the queue is held, e.g. a Workload finishes or is evicted, rather than from missing periodic samples alone. Could we align the release note/comment and integration test with that case?

I also left two held-path edge cases inline, plus one follow-up scope question.

**Comment by mimowo** [comment]
@shashankvarma49 do you have cycles to address the comments?
